MerchantValidationEvent's validationURL should use 'completeURL' to resolve document baseURL

Ahmad-S792 opened the pull request:
#### 2c886037b3061f1b3dbe86fd959608645af9b132
<pre>
MerchantValidationEvent&apos;s validationURL should use &apos;completeURL&apos; to resolve document baseURL

<a href="https://bugs.webkit.org/show_bug.cgi?id=216530">https://bugs.webkit.org/show_bug.cgi?id=216530</a>
<a href="https://rdar.apple.com/problem/69023893">rdar://problem/69023893</a>

Reviewed by Anne van Kesteren.

This patch is to use completeURL() from Document in MerchantValidationEvent&apos;s  validationURL, so it
can be resolved against document baseURL [1]. By leveraging completeURL(), we achieve this
and also align with CSS and HTML URL Parsing standards (i.e. forcing UTF8) and make this future proof.

[1] <a href="https://w3c.github.io/merchant-validation/#merchantvalidationevent-constructor">https://w3c.github.io/merchant-validation/#merchantvalidationevent-constructor</a>

&quot;Let validationURL be the result of URL parsing eventInitDict.validationURL and base.&quot;

Additionally, we didn&apos;t had the WPT test cases imported, so this PR import them from upstream.

Upstream Hash: <a href="https://github.com/web-platform-tests/wpt/commit/adc1ac02e6d8e080508325ab4525bab8d09177a6">https://github.com/web-platform-tests/wpt/commit/adc1ac02e6d8e080508325ab4525bab8d09177a6</a>

&gt; Change:
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
(MerchantValidationEvent::create):

&gt; WPT Test Cases &amp; Test Expectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/META.yml:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/complete-method.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/complete-method.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.http.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.http-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/onmerchantvalidation-attribute.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/onmerchantvalidation-attribute.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Add Platform Specific Expectations
* LayoutTests/platform/mac-wk1/TestExpectations: Ditto

&gt; Remove old imported WPT tests:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/complete-method.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/complete-method.https.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.http-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.http.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.https.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/w3c-import.log:

Canonical link: <a href="https://commits.webkit.org/273626@main">https://commits.webkit.org/273626@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5df05fdf8274e68a513fb42293ebf1d0d7d4eac1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38856 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17520 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12110 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31179 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11165 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32237 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40102 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32801 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37127 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11384 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35213 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31894 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8213 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11861 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12221 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
```diff
@@ -43,7 +43,7 @@ Ref<MerchantValidationEvent> MerchantValidationEvent::create(const AtomString& t
 ExceptionOr<Ref<MerchantValidationEvent>> MerchantValidationEvent::create(Document& document, const AtomString& type, Init&& eventInit)
 {
-    URL validationURL { document.url(), eventInit.validationURL };
+    URL validationURL { document.completeURL(eventInit.validationURL) };
```
I think this style ends up calling the URL constructor redundantly, so I'd suggest:
```suggestion
    auto validationURL = document.completeURL(eventInit.validationURL);
```

Ahmad-S792 replied:
Done.

annevk commented on Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
```diff
@@ -43,7 +43,7 @@ Ref<MerchantValidationEvent> MerchantValidationEvent::create(const AtomString& t
 ExceptionOr<Ref<MerchantValidationEvent>> MerchantValidationEvent::create(Document& document, const AtomString& type, Init&& eventInit)
 {
-    URL validationURL { document.url(), eventInit.validationURL };
+    auto validationURL = document.completeURL(eventInit.validationURL);
```
To ensure UTF-8 you need to pass ForceUTF8::Yes. Thankfully you had included the need in your commit message as I had forgotten about it.

Ahmad-S792 replied:
```auto validationURL = document.completeURL(eventInit.validationURL, ForceUTF8::Yes);```

Or I have to add `ScriptExecutionContext` as well?

annevk replied:
Yeah maybe.

Ahmad-S792 replied:
Thanks! I was under impression that completeURL() would implicitly will have UTF8 support. Thanks for correcting myself and learnt new stuff as well.

Ahmad-S792 replied:
Done in [b0a06fa](https://github.com/WebKit/WebKit/commit/b0a06fa23c984a51c24687f122f9ed1551f88014). Confirmed that it compiles locally.

annevk reviewed (approved):
Looks good modulo bots, thanks!

Ahmad-S792 committed:
MerchantValidationEvent's validationURL should use 'completeURL' to resolve document baseURL

https://bugs.webkit.org/show_bug.cgi?id=216530
rdar://problem/69023893

Reviewed by Anne van Kesteren.

This patch is to use completeURL() from Document in MerchantValidationEvent's  validationURL, so it
can be resolved against document baseURL [1]. By leveraging completeURL(), we achieve this
and also align with CSS and HTML URL Parsing standards (i.e. forcing UTF8) and make this future proof.

[1] https://w3c.github.io/merchant-validation/#merchantvalidationevent-constructor

"Let validationURL be the result of URL parsing eventInitDict.validationURL and base."

Additionally, we didn't had the WPT test cases imported, so this PR import them from upstream.

Upstream Hash: https://github.com/web-platform-tests/wpt/commit/adc1ac02e6d8e080508325ab4525bab8d09177a6

> Change:
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
(MerchantValidationEvent::create):

> WPT Test Cases & Test Expectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/META.yml:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/complete-method.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/complete-method.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.http.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.http-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/constructor.tentative.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/onmerchantvalidation-attribute.https.html:
* LayoutTests/imported/w3c/web-platform-tests/merchant-validation/onmerchantvalidation-attribute.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Add Platform Specific Expectations
* LayoutTests/platform/mac-wk1/TestExpectations: Ditto

> Remove old imported WPT tests:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/complete-method.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/complete-method.https.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.http-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.http.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/constructor.https.html:
* LayoutTests/imported/w3c/web-platform-tests/payment-request/MerchantValidationEvent/w3c-import.log:

Canonical link: https://commits.webkit.org/273626@main
